src/drivers/intel/cavs/hda.c

```diff
@@ -35,6 +43,106 @@ static int hda_get_fifo(struct dai *dai, int direction, int stream_id)
 	return 0;
 }
 
+/* Functions for HW timestamp */
+
+static inline uint32_t hda_ts_local_tsctrl_addr(void)
+{
+	return TIMESTAMP_BASE + TS_HDA_LOCAL_TSCTRL;
+}
+
+static inline uint32_t hda_ts_local_offs_addr(void)
+{
+	return TIMESTAMP_BASE + TS_HDA_LOCAL_OFFS;
+}
+
+static inline uint32_t hda_ts_local_sample_addr(void)
+{
+	return TIMESTAMP_BASE + TS_HDA_LOCAL_SAMPLE;
+}
+
+static inline uint32_t hda_ts_local_walclk_addr(void)
+{
+	return TIMESTAMP_BASE + TS_HDA_LOCAL_WALCLK;
+}
+
+static inline uint32_t hda_ts_tscc_addr(void)
+{
+	return TIMESTAMP_BASE + TS_HDA_TSCC;
+}
+
+static int hda_ts_config(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	int i;
+
+	if (cfg->type != SOF_DAI_INTEL_HDA) {
+		trace_hda_error("dmic_ts_config(): Illegal DAI type");
+		return -EINVAL;
+	}
+
+	cfg->walclk_rate = 0;
+	for (i = 0; i < NUM_SSP_FREQ; i++) {
+		if (ssp_freq_sources[i] == SSP_CLOCK_XTAL_OSCILLATOR)
+			cfg->walclk_rate = ssp_freq[i].freq;
+	}
+	trace_hda("ioclk = %d", cfg->walclk_rate);
+
+	return 0;
+}
+
+static int hda_ts_start(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	/* Set HDA timestamp registers */
+	uint32_t cdmas;
+	uint32_t addr = hda_ts_local_tsctrl_addr();
+
+	/* Set DMIC timestamp registers */
+	spin_lock(dai->lock);
+
+	/* Set CDMAS(4:0) to match DMA engine index and direction */
+	cdmas = TS_LOCAL_TSCTRL_CDMAS(cfg->dma_id |
+		(cfg->direction == SOF_IPC_STREAM_PLAYBACK ? BIT(4) : 0));
```

Comment:
@tlauda I'd need your help here. The sample count register remains zero for HDA in this version. I wonder if it could be the LLP not enabled for HDA DMA? Or just wrong DMA id here. It is also something to discuss if I should drop the HDA and DMIC part since for now at least they are not slave interfaces and should not drift vs. FW schedule timer.

The SSP and DMIC timestamps retrieve works perfectly, exactly as with settings you recommended for CDMAS.

Reply:
This is still open.

Reply:
@tlauda ping

Reply:
@singalsu @lgirdwood I'm pretty sure it should work. I can look at it as soon as I'm finished with my current effort.

Reply:
@tlauda I just re-run test with hda (topology sof-hda-asrc-2ch.tplg) and I get this error to trace `asrc.c:709 	asrc_control_loop(), DAI timestamp failed`. The time stamping programming sequence change for SSP DAI freeze as expected didn't improve things with HDA. I tried again to invert bit4 of CDMAS bit field here but no change.

When testing with playback the function hda_ts_start() seems to set every time CDMAS to value 0x17. The "1" is for playback direction and "7" for dma_id. The trace shows from elsewhere `hda-dmac: 7 channel 0 -> enable`. But I'm not sure this is the correct value to use here since it was more complicated for SSP.

Reply:
@singalsu I can successfully get samples when triggering timestamp. For HDA both directions should be 0 based in CDMAS, so something like this:
Link Input --> 0x0 for channel 0, 0x1 for channel 1, 0x2 for channel 2 etc.
Link Output --> 0x10 for channel 0, 0x11 for channel 1, 0x12 for channel 2 etc.

Reply:
Thanks!! Yes it works if I use the channel # here.

Reply:
I just pushed a new PR version with this fix. Now the DAI tracking works for HDA too. I noticed that when I attempt to play 44100 via ASRC to 48k HDA there's error:

```
   0      1          DMA         159132665.104167        12.916667 intel/cavs/hda-dma.c:653 	hda-dmac: 7 chan 0 - buffer not DMA aligned 0x2d0
    0      1         PIPE         159132689.114583        24.010416 src/audio/pipeline.c:503 	pipeline_prepare() error: ret = -22,dev->comp.id = 1
    0      1          IPC         159132708.281250        19.166666    src/ipc/handler.c:305 	ipc: pipe 1 comp 1 prepare failed -22
```
Maybe the possible odd number of frames produced causes it? I don't get this error with 48 kHz family rates.

Reply:
@tlauda @mmaka1 The same ASRC pipeline macro runs OK with 44100 input in APL and I2S codec. I don't think the ASRC should have responsibility to ensure output size to PCM buffer is N * 32 bytes. The error above looks like HDA codec driver issue to me.